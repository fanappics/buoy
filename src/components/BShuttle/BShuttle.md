```js

const options = [{id: 1, displayText: 'This'}, {id: 2, displayText: 'Is'}, {id: 3, displayText: 'An Example'}]
const selected1 = []
const selected2 = [1]
const placeholder = 'You need to do something first'

<div>
  <div>
    <b-shuttle
      :placeholder="placeholder"
      chosenLabel='These have been chosen'
      availableLabel='These can be chosen'

    />
  </div>

  <div>
    <p>Selected:</p>
  </div>
  
  <div>
    <b-shuttle
      :options="options"
      chosenLabel='These have been chosen'
      availableLabel='These can be chosen'
      v-model="selected1"
    />
  </div>

  <div>
    <p>Selected: {{ selected1 }}</p>
  </div>
  
  <div>
    <b-shuttle
      :options="options"
      chosenLabel='These have been chosen'
      availableLabel='These can be chosen'
      v-model="selected2"
    />
  </div>

  <div>
    <p>Selected: {{ selected2 }}</p>
  </div>

</div>
```