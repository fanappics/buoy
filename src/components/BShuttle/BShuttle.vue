<template>
  <div class='shuttle'>

  	<div class='available'>
  		<label id='available-label'>{{ availableLabel }}</label>
  		<ul
  			aria-labeledby='available-label'
  			role='selectbox'
  		>
  			<li
  				v-if="placeholder && (!options || options.length === 0)"
  				:aria-label="placeholder"
  				class='placeholder'
  			>
  			{{ placeholder }}
  			</li>
  			<li
  				v-for="option in availableOptions"
  				:key="'available-option-' + option.id"
  				:id="'available-option-' + option.id"
  				:class="optionsClass(option, 'available')"
  				@click="onOptionClick(option, 'available', 'chosen', $event)"
  				@keyup.space.prevent.stop="onOptionClick(option, 'available', 'chosen', $event)"
  				@keyup.down.prevent.stop="onKeyupDown($event, option, 'available', 'chosen')"
  				@keyup.up.prevent.stop="onKeyupUp($event, option, 'available', 'chosen')"
  				@keyup.16.prevent.stop="onOptionClick(option, 'available', 'chosen', $event)"
  				@keyup.right.prevent.stop="onMoveSelectedClick(selectedOptions, 'available')"
  				@keyup.enter.prevent.stop="onMoveSelectedClick(selectedOptions, 'available')"
  				@keydown.up.prevent.stop
  				@keydown.down.prevent.stop
  				@keydown.space.prevent.stop
  				role='option'
  				tabindex='0'
  			>
  			{{ option.displayText }}
  			</li>
  		</ul>
  	</div>

  	<div class='buttons'>
	  	<button 
	  		@click="onMoveAllOptions('available')"
	  		id='all-to-chosen'
	  		type='button'
	  		>
	  		All <i class='fa fa-arrow-right' aria-hidden='true'></i>
	  	</button>
	  	<button 
	  		@click="onMoveSelectedClick(selectedOptions, 'available')"
	  		id='selected-to-chosen'
	  		type='button'
	  		>
	  		<i class='fa fa-arrow-right' aria-hidden='true'></i>
	  	</button>
	  	<button
	  		@click="onMoveSelectedClick(selectedOptions, 'chosen')"
	  		id='selected-to-available'
	  		type='button'>
	  		<i class='fa fa-arrow-left' aria-hidden='true'></i>
	  	</button>
	  	<button 
	  		@click="onMoveAllOptions('chosen')"
	  		id='all-to-available'
	  		type='button'
	  		>
	  		All <i class='fa fa-arrow-left' aria-hidden='true'></i>
	  	</button>
  	</div>

  	<div class='chosen'>
  		<label id='chosen-label'>{{ chosenLabel }}</label>
  		<ul
  			aria-labeledby='chosen-label'
  			role='selectbox'
  		>
	  		<li
  				v-for="option in chosenOptions"
  				:key="'chosen-option-' + option.id"
  				:id="'chosen-option-' + option.id"
  				:class="optionsClass(option, 'chosen')"
  				@click="onOptionClick(option, 'chosen', 'available', $event)"
  				@keyup.space.stop.prevent.stop="onOptionClick(option, 'chosen', 'available', $event)"
  				@keyup.down.prevent.stop="onKeyupDown($event, option, 'chosen', 'available')"
  				@keyup.up.prevent.stop="onKeyupUp($event, option, 'chosen', 'available')"
  				@keyup.16.prevent.stop="onOptionClick(option, 'chosen', 'available', $event)"
  				@keyup.left.prevent.stop="onMoveSelectedClick(selectedOptions, 'chosen')"
  				@keyup.enter.prevent.stop="onMoveSelectedClick(selectedOptions, 'chosen')"
  				@keydown.up.prevent.stop
  				@keydown.down.prevent.stop
  				@keydown.space.prevent.stop
  				role='option'
  				tabindex='0'
  			>
  				{{ option.displayText }}
  			</li>
  		</ul>
  	</div>

  </div>
  
</template>

<script>
export default {
  name: 'b-shuttle',
  props: {
  		/**
  		* The available options for the shuttle.
  		*/
  	options: {
  		type: Array,
  		required: false
  	},
  		/**
  		* Placeholder to display if shuttle is dependent on other component.
  		*/
  	placeholder: {
  		type: String,
  		required: false
  	},
  		/**
  		* Label to display above available options.
  		*/
  	availableLabel: {
  		type: String,
  		required: true
  	},
  		/**
  		* Label to display above chosen options.
  		*/
  	chosenLabel: {
  		type: String,
  		required: true
  	},
    /**
    * The id of the selected option.
    */
    value: {
      type: Array,
      required: false,
      default: function () { return [] }
    }
  },

  data () {
    return {
    	availableOptions: (this.options && this.options.length > 0) ? this.setOptions(this.options, this.value, 'available') : new Array,
    	selectedOptions: {'available': new Array, 'chosen': new Array},
    	chosenOptions: (this.value && this.options) ? this.setOptions(this.options, this.value, 'chosen') : new Array,
    	lastClick: 0
    }
  },

  computed: {
  },

  methods: {
  	setOptions: function (options, selected, type) {
  		const availableOptions = new Array
  		const chosenOptions = new Array
  		options.forEach(function (option) {
  			if (selected === null) {
  				availableOptions.push(option)
  			} else if (selected.indexOf(option.id) === -1) {
  				availableOptions.push(option)
  			} else {
  				chosenOptions.push(option)
  			}
  		})
  		if (type === 'available') {
	      return this.sortById(availableOptions)
  		} else if (type === 'chosen') {
	      return this.sortById(chosenOptions)  	
  		}
    },
  	optionsClass: function (option, optionType) {
      return {
        'options': true,
        'selected': this.selectedOptions[optionType].indexOf(option.id) > -1
      }
    },
  	onOptionClick (option, optionType, otherOptionType, event) {
  		if (this.selectedOptions[otherOptionType].length > 0) { this.selectedOptions[otherOptionType] = new Array }
  		if (event.shiftKey && (this.lastClick > 0)) {
  			this.multipleSelect(option, optionType)
  		}
  		this.selectOption(option.id, optionType)
  	},
  	onMoveSelectedClick (options, type) {
  		const loopOptions = (type === 'available') ? this.availableOptions : this.chosenOptions
  		const availableOptions = new Array
  		const chosenOptions = new Array
  		loopOptions.forEach((option) => {
  			const index = this.selectedOptions[type].indexOf(option.id)
  			if (index > -1) {
  				(type === 'available') ? chosenOptions.push(option) : availableOptions.push(option)
  			} else {
  				(type === 'available') ? availableOptions.push(option) : chosenOptions.push(option)
  			}
  		})
  		if (type === 'available') {
  			this.availableOptions = this.sortById(availableOptions)
  			this.chosenOptions.push.apply(this.chosenOptions, chosenOptions)
  			this.sortById(this.chosenOptions)
  		} else {
  			this.availableOptions.push.apply(this.availableOptions, availableOptions)
  			this.sortById(this.availableOptions)
  			this.chosenOptions = this.sortById(chosenOptions)
  		}
  		this.selectedOptions = {'available': new Array, 'chosen': new Array}
  		this.$emit("input",Array.from(this.chosenOptions, option => option.id));
  	},
  	onMoveAllOptions (type) {
  		if (type === 'available') {
  			this.chosenOptions = this.sortById(this.options)
  			this.availableOptions = new Array
  		} else if (type === 'chosen') {
  			this.chosenOptions = new Array
  			this.availableOptions = this.sortById(this.options)
  		}
  		this.selectedOptions = {'available': new Array, 'chosen': new Array}
  		this.$emit("input",Array.from(this.chosenOptions, option => option.id));
  	},
     /**
     * Handles the down arrow (40) keyup event
     */
    onKeyupDown (event, option, optionType, otherOptionType) {
      const target = event.target;
      if(target.nextElementSibling) {
        const next = target.nextElementSibling;
        next.focus();
  			if (event.shiftKey) {
        	if (this.selectedOptions[otherOptionType].length > 0) { this.selectedOptions[otherOptionType] = new Array }
  				this.selectOption(option.id, optionType)
  			}
      }
    },
    /**
     * Handles the up arrow (38) keyup event
     */
    onKeyupUp (event, option, optionType, otherOptionType) {
      const target = event.target;
      if(target.previousElementSibling) {
        const next = target.previousElementSibling;
        next.focus();
  			if (event.shiftKey) {
        	if (this.selectedOptions[otherOptionType].length > 0) { this.selectedOptions[otherOptionType] = new Array }
  				this.selectOption(option.id, optionType)
  			}      }
    }, 
  	selectOption (id, optionType) {
  		const index = this.selectedOptions[optionType].indexOf(id)
  		if (index > -1) {
  			this.selectedOptions[optionType].splice(index, 1)
  		} else {
  			this.selectedOptions[optionType].push(id)
  		}
  		this.lastClick = id
  	},
  	/**
  	 * Expected behavior here is that a preselected item that is hit with a multi-select
  	 * will flip its state accordingly.
  	 */
  	multipleSelect (option, optionType) {
  		const high = option.id > this.lastClick ? option.id : this.lastClick
  		const low = option.id > this.lastClick ? this.lastClick : option.id
  		const multipleOptions = this.options
  			.filter(item => (item.id > low && item.id < high))
  			.map(function(obj){return obj.id})
  		
  		multipleOptions.forEach( (optionId) => {
  			const index = this.selectedOptions[optionType].indexOf(optionId)
  			this.selectOption(optionId, optionType)
  		})
  	},
  	sortById (options) {
  		options.sort(function (itemOne, itemTwo) {
  			return itemOne.id - itemTwo.id
  		})
      return options 
  	}
  }
}
</script>

<style scoped>

	ul {
		list-style: none;
		border-style: solid;
		border-width: thin;
		border-color: #dededf;
		height: 10rem;
		padding: 0;
		margin: 0;
		overflow: scroll;
    overflow-x: hidden;
	}

	div.available {
		flex: 1;
		margin-right: 1rem;
	}

	div.chosen {
		flex: 1;
		margin-left: 1rem;
	}

	div.buttons {
		margin-top: 1.2rem
	}

	button {
		display: block;
		border-color: #dededf;
		padding: .25rem .50rem .25rem .50rem;
		margin: .3rem 0 .3rem 0;
		background: none;
		border-radius: .25rem;
		width: 3rem;
	}

	button:hover {
		cursor: pointer;
	}

	.shuttle {
		display: flex;
	}

	i {
		color: #00aaed;
	}

	li.placeholder {
		color: #dededf;
	}

	li.options:hover {
		cursor: pointer;
	}

	li.selected {
		background-color: #00aaed;
		color: #ffffff;
	}

</style>
