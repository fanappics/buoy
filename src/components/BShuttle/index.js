import BShuttle from './BShuttle.vue'

/* istanbul ignore next */
BShuttle.install = function install (Vue) {
  Vue.component(BShuttle.name, BShuttle)
}

export default BShuttle
