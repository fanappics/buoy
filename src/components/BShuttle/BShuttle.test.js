import { createLocalVue, shallow } from 'vue-test-utils'
import BShuttle from './BShuttle'

import VeeValidate from 'vee-validate'
const localVue = createLocalVue()
localVue.use(VeeValidate)

describe('BShuttle', () => {
  const shuttleOptions = [{ id: 1, displayName: 'test-1' }, { id: 2, displayName: 'test-2' }]
  const placeholder = 'Testing Placeholder'
  const availableLabel = 'test available label'
  const chosenLabel = 'test chosen label'
  const prechosen = [1]

  it('BShuttle renders correctly', () => {
    const wrapper = shallow(BShuttle, {
      propsData: {
        options: shuttleOptions,
        availableLabel: availableLabel,
        chosenLabel: chosenLabel
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('creates options when passed', () => {
    const wrapper = shallow(BShuttle, {
      propsData: {
        options: shuttleOptions,
        availableLabel: availableLabel,
        chosenLabel: chosenLabel
      },
      localVue
    })
    expect(wrapper.props().options.length).toBe(shuttleOptions.length)
  })

  it('displays placeholder if no initial options', () => {
    const wrapper = shallow(BShuttle, {
      propsData: {
        availableLabel: availableLabel,
        chosenLabel: chosenLabel,
        placeholder: placeholder
      },
      localVue
    })
    const div = wrapper.find('li')
    expect(div.text()).toBe(placeholder)
  })

  it('intializes chosen value when passed', () => {
    const wrapper = shallow(BShuttle, {
      propsData: {
        options: shuttleOptions,
        availableLabel: availableLabel,
        chosenLabel: chosenLabel,
        value: prechosen
      },
      localVue
    })
    expect(wrapper.vm.chosenOptions.length).toBe(1)
    expect(wrapper.vm.chosenOptions[0].id).toBe(1)
  })

  it('highlights row when clicked', async () => {
    const wrapper = shallow(BShuttle, {
      propsData: {
        options: shuttleOptions,
        availableLabel: availableLabel,
        chosenLabel: chosenLabel
      },
      localVue
    })
    await wrapper.vm.$nextTick()
    const line = wrapper.find(`#available-option-${wrapper.vm.id}-${shuttleOptions[0].id}`)
    expect(line.classes()).not.toContain('selected')
    line.trigger('click')
    expect(line.classes()).toContain('selected')
  })

  it('moves row when chosen arrow clicked', async () => {
    const wrapper = shallow(BShuttle, {
      propsData: {
        options: shuttleOptions,
        availableLabel: availableLabel,
        chosenLabel: chosenLabel
      },
      localVue
    })
    await wrapper.vm.$nextTick()
    const line = wrapper.find(`#available-option-${wrapper.vm.id}-${shuttleOptions[0].id}`)
    const button = wrapper.find('#selected-to-chosen')
    line.trigger('click')
    button.trigger('click')
    expect(wrapper.vm.chosenOptions.length).toBe(1)
    expect(wrapper.vm.chosenOptions[0].id).toBe(1)
  })

  it('moves row when available arrow clicked', async () => {
    const wrapper = shallow(BShuttle, {
      propsData: {
        options: shuttleOptions,
        availableLabel: availableLabel,
        chosenLabel: chosenLabel
      },
      localVue
    })
    await wrapper.vm.$nextTick()
    const line = wrapper.find(`#available-option-${wrapper.vm.id}-${shuttleOptions[0].id}`)
    const button = wrapper.find('#selected-to-chosen')
    line.trigger('click')
    button.trigger('click')
    const chosenLine = wrapper.find(`#chosen-option-${wrapper.vm.id}-${shuttleOptions[0].id}`)
    const chosenButton = wrapper.find('#selected-to-available')
    chosenLine.trigger('click')
    chosenButton.trigger('click')
    expect(wrapper.vm.chosenOptions.length).toBe(0)
    expect(wrapper.vm.availableOptions.length).toBe(2)
    expect(wrapper.vm.availableOptions[0].id).toBe(1)
  })

  it('moves all rows to chosen when button clicked', () => {
    const wrapper = shallow(BShuttle, {
      propsData: {
        options: shuttleOptions,
        availableLabel: availableLabel,
        chosenLabel: chosenLabel
      },
      localVue
    })
    const button = wrapper.find('#all-to-chosen')
    expect(wrapper.vm.chosenOptions.length).toBe(0)
    button.trigger('click')
    expect(wrapper.vm.chosenOptions.length).toBe(2)
  })

  it('moves all rows to available when button clicked', () => {
    const wrapper = shallow(BShuttle, {
      propsData: {
        options: shuttleOptions,
        availableLabel: availableLabel,
        chosenLabel: chosenLabel
      },
      localVue
    })
    const chosenButton = wrapper.find('#all-to-chosen')
    const availableButton = wrapper.find('#all-to-available')
    expect(wrapper.vm.chosenOptions.length).toBe(0)
    chosenButton.trigger('click')
    expect(wrapper.vm.chosenOptions.length).toBe(2)
    expect(wrapper.vm.availableOptions.length).toBe(0)
    availableButton.trigger('click')
    expect(wrapper.vm.chosenOptions.length).toBe(0)
    expect(wrapper.vm.availableOptions.length).toBe(2)
  })
})
