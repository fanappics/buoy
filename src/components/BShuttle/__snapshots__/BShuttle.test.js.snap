// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BShuttle BShuttle renders correctly 1`] = `
<div class="shuttle">
    <div class="b-shuttle-available">
        <label id="available-label-null">test available label</label>
        <ul aria-labelledby="available-label-null" tabindex="-1" role="listbox" class="input">
            <!---->
            <li id="available-option-null-1" role="option" tabindex="0" class="b-shuttle-options">

            </li>
            <li id="available-option-null-2" role="option" tabindex="0" class="b-shuttle-options">

            </li>
        </ul>
    </div>
    <div class="b-shuttle-buttons">
        <button aria-label="Move All To Chosen" id="all-to-chosen" type="button">
            <img src="[object Object]" class="icon">
        </button>
        <button aria-label="Move Selected To Chosen" id="selected-to-chosen" type="button">
            <i aria-hidden="true" class="icon ion-arrow-right-c"></i>
        </button>
        <button aria-label="Move Selected To Available" id="selected-to-available" type="button">
            <i aria-hidden="true" class="icon ion-arrow-left-c"></i>
        </button>
        <button aria-label="Move All To Available" id="all-to-available" type="button">
            <img src="[object Object]" class="icon">
        </button>
    </div>
    <div class="b-shuttle-chosen">
        <label id="chosen-label-null">test chosen label</label>
        <ul aria-labelledby="chosen-label-null" role="listbox" class="input"></ul> <span id="error-null" class="error-text" style="display: none;"></span></div>
</div>
`;
